firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/chunks.txt

```diff
@@ -0,0 +1,152 @@
+      THEN("The input method reports output_ready status")
+      THEN("the output method reports ok status")
+      THEN("The output buffer is as expected")
+
+Scenario: After output is consumed, ChunkSplitter's behavior resets to that of a ChunkSplitter with an empty buffer
+
+  GIVEN("A uint8_t ChunkSplitter completely filled with input data, and no delimiters are consumed with an output call")
+    WHEN("10 bytes of non-delimiter bytes followed by a delimiter byte are passed as input, and output is called between each input")
+      THEN("The input method reports ok status")
+      THEN("The output method reports waiting status")
+      THEN("The input method reports output_ready status")
+      THEN("The output method reports ok status")
+      THEN("The output buffer is as expected")
+
+Scenario: chunksplitter correctly handles delimeters other than 0x00
+
+  GIVEN("A uint8_t chunksplitter with a completely full internal buffer of capacity 256 bytes, with no delimiters which is equal to 0x01")
+    WHEN("The output method is called")
+      THEN("The output method returns waiting status")
+      THEN("The output buffer is empty")
+
+    WHEN("delimeter 0x01 is passed as input and output is called")
+      THEN("the input method reports output ready status")
+      THEN("the output status is equal to ok")
+      THEN("The output buffer is as expected")
+
+  GIVEN("A uint8_t chunksplitter with a completely full internal buffer of capacity 256 bytes, with no delimiters which is equal to 0xff")
+    WHEN("The Output method is called")
+      THEN("The output method returns waiting status")
+
+    WHEN("delimeter 0xff is passed as input and output is called")
+      THEN("the input method reports output ready status")
+      THEN("the output status is equal to ok")
+      THEN("The output buffer is as expected")
+
+  GIVEN("A uint8_t chunksplitter with a completely full internal buffer of capacity 256 bytes, with include_delimeter bool equal to true")
+    WHEN("The output method is called")
+      THEN("the output method reports ok status")
+      THEN("The output buffer is as expected")
+
+    WHEN("delimeter 0x01 is passed as input and output is called")
+      THEN("the input method reports invalid_length status") // as the buffer is full
+      THEN("the output method reports invalid length status")
+      THEN("The output buffer is as expected")
+
+  GIVEN("A uint8_t chunksplitter with a partially filled internal buffer of capacity 256 bytes, with include_delimeter bool equal to true")
+    WHEN("The output method is called")
+      THEN("the output method reports ok status")
+
+    WHEN("Output is called after passing delimiter equal to 0x01 as input")
+      THEN("The input method reports ok status")
+      THEN("the output method reports ok status")
+      THEN("The output buffer contains delimiter")
+
+Scenario: Chunksplitter correctly handles input of multiple delimited chunks before output consumes them
```

Comment:
Some notes on invariants and scenarios for the desired behavior:
- when we add more inputs without consuming outputs, we discard the previous outputs and just store the new inputs as the current chunk
- the first byte added after an unconsumed output resets the chunksplitter's behavior to be identical to the first byte added after a consumed output; except for the include_delimiter flag
- the internal buffer of chunksplitter always stores between 0 and 1 chunks
- scenario: include_delimiter controls whether the delimiter byte is the last byte of the output, but does not affect anything